File: 2872-maximum-number-of-k-divisible-components/2872-maximum-number-of-k-divisible-components.cpp
class Solution {
public:
    vector<vector<int>> adj;
    int ans = 0;
    vector<bool>vis;
    int dfs(int node,vector<int>& values, int k){
        vis[node] = true;
        long long sum = values[node];
        for(int child : adj[node]){
            if(vis[child])continue;
            sum+=dfs(child,values,k);
            sum%=k;
        }
        if(sum%k==0)ans++;
        return sum;
    }
    
    int maxKDivisibleComponents(int n, vector<vector<int>>& edges, vector<int>& values, int k) {
        if(n==1)return 1;
        adj.resize(n);
        vis.resize(n);
        vector<int>degree(n);
        for(auto it : edges){
            adj[it[0]].push_back(it[1]);
            adj[it[1]].push_back(it[0]);
            degree[it[0]]++;
            degree[it[1]]++;
        }
        queue<int>q;
        for(int i = 0 ; i < n;i++)if(degree[i]==1)q.push(i);
        while(q.size()){
            int node = q.front();q.pop();
            degree[node]--;
            if(values[node]%k==0)ans++;

            for(int child : adj[node]){
                degree[child]--;
                values[child]+=values[node];
                values[child]%=k;
                if(degree[child]==1)q.push(child);
            }
        }

        return ans;

    }
};

File: 2872-maximum-number-of-k-divisible-components/2872-maximum-number-of-k-divisible-components.py
from collections import defaultdict, deque

class Solution:
    def maxKDivisibleComponents(self, n: int, edges: list[list[int]], values: list[int], k: int) -> int:
        if n == 1:
            return 1
        
        adj = defaultdict(list)
        degree = [0] * n
        for u, v in edges:
            adj[u].append(v)
            adj[v].append(u)
            degree[u] += 1
            degree[v] += 1
        
        q = deque()
        for i in range(n):
            if degree[i] == 1:
                q.append(i)
        
        ans = 0
        while q:
            node = q.popleft()
            degree[node] -= 1
            if values[node] % k == 0:
                ans += 1
            
            for child in adj[node]:
                degree[child] -= 1
                values[child] += values[node]
                values[child] %= k
                if degree[child] == 1:
                    q.append(child)
        
        return ans
